### DISCUSSION TITLE
How to map Set -> List

### DISCUSSION THREAD

**Discussion by dave-at-season** [ExpediaGroup/graphql-kotlin#1110] [Q&A] [answered]
Hello! First off, I understand the reasoning for not supporting the automatic mapping of Kotlin Sets to GraphQL List types as per the docs and the various issue/pr conversations around the topic. 

But I am trying to add graphql to an existing Kotlin project that has a collection of domain objects defined as data classes which heavily use Sets. So I will need some way to do this and its not feasible for me to do this mapping manually. Currently I'm trying to do that by maintaining separate copies of these data classes and mapping all of their properties manually. For example:

```kotlin
data class DomainClient(
    val firstName: String,
    val lastName: String,
    // ... many more properties,
    val locations: Set<Location> = emptySet()
)
data class Client(
    val firstName: String,
    val lastName: String,
    // ... many more properties,
    val locations: List<Location> = emptyList()
)
fun clientFromDomain(c: DomainClient): Client = Client(
    firstName = c.firstName,
    lastName = c.lastName,
    // ... many more properties,
    locations = c.locations.toList()
)

@Provider
class ClientsQuery @Inject constructor(
    private val clientService: ClientService
) : Query {
    fun clients(context: CustomGraphQLContext): List<Client> {
        return clientService.getClients(context.user.accountId)
            .map { clientFromDomain(it) }
    }
}
```
Which is exceedingly yucky given the number of classes/fields I need to map across this project.

All I really want is a way to automate this mapping, I don't even necessarily want that reflected in the resulting schema. I'm fine with the schema outputting these fields as a List type, and then serializing/deserializing these fields into Sets behind the scenes as long as I don't have to do it manually.

It seems like functionality was introduced to accommodate this workflow as discussed in this issue: https://github.com/ExpediaGroup/graphql-kotlin/issues/584 but I don't understand how `generateAdditionalTypes` would be used here.

I would greatly appreciate any help on this

**Comment by dariuszkuc** [comment, marked as answer]
Hello 👋 
I think the "simplest" way would be to follow the suggestion from #584 with small update -> with latest code you can just pass the additional types to the generator directly.

1. Create custom `SchemaGeneratorHooks`

```kotlin
override fun willGenerateGraphQLType(type: KType): GraphQLType? = when (type.classifier) {
  Set::class -> GraphQLList.list(GraphQLTypeReference.typeRef(type.getTypeOfFirstArgument().getSimpleName()))
  else -> super.willGenerateGraphQLType(type)
}

override fun isValidSuperclass(kClass: KClass<*>): Boolean {
  return if(kClass == Set::class) false else  super.isValidSuperclass(kClass)
}
```

2. Generate schema by explicitly passing additional `Set` types to ensure they get generated

```kotlin
val additionalTypes = mutableSetOf(MyClass::class.createType())
val generator = SchemaGenerator(config) // config should use your custom hooks from above
return generator.use {
    it.generateSchema(queries, mutations, subscriptions, additionalTypes)
}
```

**Reply by smyrick** [reply]
I think I have a slightly better approach:

Update the `willResolveMonad` hooks

```kt
override fun willResolveMonad(type: KType): KType = when (type.classifier) {
    Set::class -> List::class.createType(type.arguments)
    else -> type
}
```


Write your query as normal

```kt
fun setList(): Set<String> = setOf("one", "one", "two")
```


Results are returned unique

![Screen Shot 2021-04-02 at 2 12 12 PM](https://user-images.githubusercontent.com/2446877/113454643-900ff900-93bd-11eb-80a9-48fcdce5897d.png)

**Reply by dariuszkuc** [reply]
Nice!

**Reply by dave-at-season** [reply]
Thanks for the suggestions! This got me closer.

I'm probably missing something, but I think that won't work if the Set is a non-function property on a data class. Since `willResolveMonad` is only called by `generateFunction` and not `generateProperty`

Full Example:

```kotlin
data class SetList(
    val city: String,
    val songs: Set<String>
)

class SetListQuery : Query {
    fun setList(): SetList = SetList(city = "Philadelphia, PA", songs = setOf("one", "one", "two"))
}

class CustomGeneratorHooks : SchemaGeneratorHooks {
    override fun willResolveMonad(type: KType): KType = when (type.classifier) {
        Set::class -> List::class.createType(type.arguments)
        else -> type
    }
}

val config = SchemaGeneratorConfig(
    supportedPackages = listOf(
        "...",
        "kotlin.collections"
    ),
    hooks = CustomGeneratorHooks(),
)

val schema: GraphQLSchema = toSchema(
        config = config,
        queries = listOf(
            TopLevelObject(recipeQuery),
            TopLevelObject(recipesQuery),
            TopLevelObject(ingredientQuery),
            TopLevelObject(clientsQuery),
            TopLevelObject(SetListQuery())
        ),
        mutations = listOf(TopLevelObject(updateRecipeStatusMutation)),
)
```

**Reply by dave-at-season** [reply]
I feel like what I really want here is some way of doing something like this:

```kotlin
class CustomGeneratorHooks : SchemaGeneratorHooks {
    override fun willGenerateGraphQLType(type: KType): GraphQLType? = when (type.classifier) {
        // made up function that returns a GraphQLType for the given KType
        Set::class -> generateGraphQLType(List::class.createType(type.arguments))
        else -> super.willGenerateGraphQLType(type)
    }
}
```
This is similar to @dariuszkuc's suggestion but would (I think) bypass the need to manually add all the wrapped types via additionalTypes. I think that would cause confusion for folks working on the app. (They add a new type and don't understand why a strange error is thrown when they try and use it in a Set)

Alternatively just some way to optionally override `isListType` to include `Set` would be nice

**Reply by dave-at-season** [reply]
This appears to work for what I need:

```kotlin
class CustomGeneratorHooks : SchemaGeneratorHooks {
    override fun willGenerateGraphQLType(type: KType): GraphQLType? = when (type.classifier) {
        Set::class -> GraphQLList.list(
            GraphQLTypeReference.typeRef(
                type.arguments.firstOrNull()?.type?.jvmErasure?.simpleName
            )
        )
        else -> super.willGenerateGraphQLType(type)
    }

    override fun isValidSuperclass(kClass: KClass<*>): Boolean {
        return if (kClass == Set::class) false else super.isValidSuperclass(kClass)
    }
}
```

**Reply by dariuszkuc** [reply]
If you do type ref then you have to ensure that underlying reference will be build (i.e. it is referenced from some other place in graph) - that is why those were explicitly added in my example.

**Reply by dave-at-season** [reply]
yep! I just noticed that trying to work this into the actual project I'm working on. It worked in this trivial example because the sets were of Strings, not generated types

**Reply by dave-at-season** [reply]
would y'all be open to a PR that allowed for optionally overriding isListType to include Set? That seems like the cleanest way to accomplish this to me

**Reply by smyrick** [reply]
@dave-at-season See the discussion here on why we don't support Set as a collection: https://github.com/ExpediaGroup/graphql-kotlin/issues/201

What if instead we updated the hooks `willResolveMonad` to be called on both properties and functions. That would be a backwards compatible change for everyone

**Reply by dave-at-season** [reply]
@smyrick yea I think that would be great

**Comment by TiKevin83** [comment]
For anyone who got here and was still confused, the [example project](https://github.com/ExpediaGroup/graphql-kotlin/blob/053bce00d7983d6466a22b1cb5bf3ef963730051/examples/server/spring-server/src/main/kotlin/com/expediagroup/graphql/examples/server/spring/hooks/CustomSchemaGeneratorHooks.kt#L62) has the current solution

```kotlin
class CustomSchemaGeneratorHooks() : SchemaGeneratorHooks {
    override fun willResolveMonad(type: KType): KType = when (type.classifier) {
        Set::class -> List::class.createType(type.arguments)
        else -> super.willResolveMonad(type)
    }
}
```

And then to use this in a spring server, this is also needed from the [example project](https://github.com/ExpediaGroup/graphql-kotlin/blob/053bce00d7983d6466a22b1cb5bf3ef963730051/examples/server/spring-server/src/main/kotlin/com/expediagroup/graphql/examples/server/spring/Application.kt#L42):

```kotlin
@Configuration
class GraphQLConfiguration {
  @Bean
  fun hooks() = CustomSchemaGeneratorHooks()
}
```

**Reply by tyrantkhan** [reply]
@dariuszkuc @dave-at-season @TiKevin83 are any of you aware how I could employ this solution if I wrap my queries in CompleteableFutures?

The type ends up being ```java.util.concurrent.CompletableFuture<kotlin.collections.Set<kotlin.String>>```

```
@GraphQLDescription("Sample query")
    @Suppress("unused")
    fun sampleQuery(
        dfe: DataFetchingEnvironment,
    ): CompletableFuture<Set<String>> {
        return CompletableFuture.supplyAsync {
            setOf(
                "foo",
            )
        }
    }
```

Not using CFs results in the sample mechansim working. Any thoughts?

**Reply by tyrantkhan** [reply]
So I did this... but it looks pretty gnarly 

```
override fun willResolveMonad(type: KType): KType = when (type.classifier) {
        Set::class -> List::class.createType(type.arguments)
        CompletableFuture::class -> {
            val innerType = type.arguments.first().type!!
            when (innerType.classifier) {
                Set::class -> List::class.createType(innerType.arguments)
                else -> type
            }
        }
        else -> type
    }
```

**Reply by dariuszkuc** [reply]
Our current logic first looks at the value from hooks and then attempts to unwrap it if needed (i.e. from `CompletableFuture`, `DataFetcherResult`, etc) so sadly I don't think there is an easier way than your existing hook (personally I'd just update the return type to a list and keep it simple).

**Reply by smyrick** [reply]
We could possibly open up the return type parser to be recursive in hooks. We do have a recursive function today but once it hit hooks we are usually expecting it to be the terminal type.

**Reply by tyrantkhan** [reply]
Sorry to ping this discussion thread yet again. While the above worked if a return type is a set, but what if it's in the input? I normally wouldn't bother and just use lists, but i'm moving a fairly large api to using this project for schema generation -- so it would be a challenge to convert all usages of set to list.

```
fun sampleQuery(
        input: Set<String>,
        dfe: DataFetchingEnvironment,
    ): List<String> {
        return  listOf()
    }
```

Results in 
```
com.expediagroup.graphql.generator.exceptions.InvalidInputFieldTypeException: Argument cannot be an interface or a union, parameter #1 input of fun RootQueryTypeResolver.sampleQuery(kotlin.collections.Set<kotlin.String>, graphql.schema.DataFetchingEnvironment): kotlin.collections.List<kotlin.String>
```

**Reply by tyrantkhan** [reply]
ooh i see there is a hook for input monads :) Need to also create! willResolveInputMonad

**Reply by coolbharat78** [reply]
Can we support reactor Flux also using same way? I wrote a MonadHook + custom fetcher but it complains for 
Exception in thread "main" com.expediagroup.graphql.generator.exceptions.TypeNotSupportedException: Cannot convert reactor.core.publisher.Flux<com.example.Gues> since it is not a valid GraphQL type or outside the supported packages "[com.example]"


```
class MonadHooks : SchemaGeneratorHooks {
    override fun willResolveMonad(type: KType): KType = when (type.classifier) {
        Mono::class -> type.arguments.firstOrNull()?.type
        Flux::class -> List::class.createType(type.arguments)
        Flow::class -> List::class.createType(type.arguments)
        Set::class -> List::class.createType(type.arguments)
        else -> type
    } ?: type
}

class CustomFunctionDataFetcher(target: Any?, fn: KFunction<*>) : FunctionDataFetcher(target, fn) {
    override fun get(environment: DataFetchingEnvironment): Any? {
        val result = super.get(environment)
        when (result) {
            is Mono<*> -> return result.toFuture()
            is Flux<*> -> return result.log().collectList().toFuture()
            else -> return result
        }
    }
}
```

**Reply by smyrick** [reply]
The hooks should override Flux the same way as Set, just make sure the `Flux::class` you are mapping is infact the correct package and not some other Flux, like from [coroutines](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-reactor/kotlinx.coroutines.reactor/flux.html) vs [Spring Flux](https://projectreactor.io/docs/core/release/api/reactor/core/publisher/Flux.html)

**Reply by coolbharat78** [reply]
qq. How will Flux and Mono work with dataloaders? Can I return a Flux or Mono from the dataloader or I have completely misunderstood dataloaders?
override fun getDataLoader(graphQLContext: GraphQLContext): 
DataLoader<Pair<String, List<String>>, Flux< User > = 
DataLoaderFactory.newDataLoader { keys -> Flux.of(User())}

**Reply by smyrick** [reply]
I am not quite sure about that but the docs here describe how you can use Dataloader with Coroutines https://opensource.expediagroup.com/graphql-kotlin/docs/server/data-loader/#dataloaders-and-coroutines
